airflow-core/src/airflow/api_fastapi/core_api/routes/ui/grid.py

```diff
@@ -257,8 +268,324 @@ def grid_data(
         )
         for dag_run in dag_runs
     ]
+    return GridResponse(dag_runs=grid_dag_runs)
+
+
+def _get_latest_serdag(dag_id, session):
+    serdag = session.scalar(
+        select(SerializedDagModel)
+        .where(
+            SerializedDagModel.dag_id == dag_id,
+        )
+        .order_by(SerializedDagModel.id.desc())
+        .limit(1)
+    )
+    if not serdag:
+        raise HTTPException(
+            status.HTTP_404_NOT_FOUND,
+            f"Dag with id {dag_id} was not found",
+        )
+    return serdag
+
+
+def _get_serdag(dag_id, dag_version_id, session) -> SerializedDagModel | None:
+    # this is a simplification - we account for structure based on the first task
+    version = session.scalar(select(DagVersion).where(DagVersion.id == dag_version_id))
+    if not version:
+        version = session.scalar(
+            select(DagVersion)
+            .where(
+                DagVersion.dag_id == dag_id,
+            )
+            .order_by(DagVersion.id)  # ascending cus this is mostly for pre-3.0 upgrade
+            .limit(1)
+        )
+    if not (serdag := version.serialized_dag):
+        log.error(
+            "No serialized dag found",
+            dag_id=dag_id,
+            version_id=version.id,
+            version_number=version.version_number,
+        )
+    return serdag
+
+
+@grid_router.get(
+    "/structure/{dag_id}",
```

Comment:
Nit it is not too late (but also don't let this stop merging!)

```suggestion
    "/{dag_id}/structure",
```

As the structure is "a property" of the dag this feels more natural to me

Reply:
i'll do these as followup